File: 06-observability/03-correlation-ids/cpp/lose_the_context.cpp
// Layer 6 Topic 3 - Losing trace context at a C++ concurrency boundary.
//
// What this demonstrates
// ----------------------
// C++ has no ambient context of any kind. There is no contextvar, no
// AsyncLocalStorage, no ThreadLocal-backed Context object shipped with the
// language. You have exactly two options, and this file runs both:
//
//   RUN 1  Thread a context struct through every call.
//          Verbose, mechanical, and correct -- the compiler will not let you
//          forget a parameter. This is Go's answer arrived at by necessity.
//
//   RUN 2  Put the context in a `thread_local`.
//          Ergonomic, and exactly wrong the moment a thread pool reuses a
//          thread for a different request. You do not lose the context: you
//          inherit the PREVIOUS request's, which is worse than a truncated
//          trace, because it produces a complete-looking trace that is false.
//          Nothing errors. Nothing is empty. The ID is simply someone else's.
//
// This is why the shared output shape has a third verdict that the other five
// languages do not use: `WRONG (inherited from previous request)`. A truncated
// trace makes you look for the break. A wrong one makes you look at the wrong
// request, and you will not know you are doing it.
//
// The pool below has TWO worker threads, each with its own queue, and requests
// are dispatched by index -- so which thread serves which request is decided by
// this file, not by the OS scheduler. That makes run 2 reproducible instead of
// a race you might not hit. In a real pool it is a race, which is the reason
// this bug survives testing.
//
// macOS/arm64 note: this uses only <thread>, <mutex> and <condition_variable>,
// which are portable. Build with -pthread; Apple clang accepts it.
//
// What to look for in the output
// ------------------------------
// The shared shape:
//
//   caller trace_id   <id>
//   callee trace_id   <id or "none">   naive
//   callee trace_id   <id>             propagated
//   verdict           lost | preserved | WRONG (inherited from previous request)
//
// Then run 2's per-request table. Request C never set the thread-local (its
// handler was a queue drain that assumed context was "already there"), landed
// on the thread that served request A, and reported A's trace ID. Compare the
// `observed` and `truth` columns: they disagree, and no line of that program
// could have noticed.

#include <atomic>
#include <condition_variable>
#include <cstdint>
#include <cstdio>
#include <functional>
#include <mutex>
#include <optional>
#include <queue>
#include <sstream>
#include <string>
#include <thread>
#include <vector>

// ---------------------------------------------------------------------------
// A minimal span and the W3C traceparent codec.
// ---------------------------------------------------------------------------

struct Span {
    std::string name;
    std::string trace_id;
    std::string span_id;
    bool sampled = true;
};

static uint64_t xorshift_state = 0x2026081800000001ULL;

static std::string rand_hex(int bytes) {
    // Fixed-seed xorshift. Not secure and does not need to be: nothing here
    // leaves the process.
    std::ostringstream out;
    for (int i = 0; i < bytes; ++i) {
        xorshift_state ^= xorshift_state << 13;
        xorshift_state ^= xorshift_state >> 7;
        xorshift_state ^= xorshift_state << 17;
        char buf[3];
        std::snprintf(buf, sizeof(buf), "%02x",
                      static_cast<unsigned>((xorshift_state >> 24) & 0xff));
        out << buf;
    }
    return out.str();
}

static Span make_span(const std::string& name) {
    Span s;
    s.name = name;
    s.trace_id = rand_hex(16);
    s.span_id = rand_hex(8);
    return s;
}

static std::string traceparent_of(const Span& s) {
    return "00-" + s.trace_id + "-" + s.span_id + "-" + (s.sampled ? "01" : "00");
}

static std::optional<Span> span_from_traceparent(const std::string& header,
                                                 const std::string& name) {
    // version-traceid-spanid-flags
    if (header.size() != 55 || header.substr(0, 3) != "00-") return std::nullopt;
    Span s;
    s.name = name;
    s.trace_id = header.substr(3, 32);
    s.span_id = header.substr(36, 16);
    s.sampled = header.substr(53, 2) != "00";
    if (header[35] != '-' || header[52] != '-') return std::nullopt;
    return s;
}

// ---------------------------------------------------------------------------
// A fixed thread pool where placement is decided by the caller. Each worker
// owns a queue, so `submit(worker_index, task)` is deterministic.
// ---------------------------------------------------------------------------

class PinnedPool {
public:
    explicit PinnedPool(int workers) : queues_(workers), mutexes_(workers), cvs_(workers) {
        for (int i = 0; i < workers; ++i) {
            threads_.emplace_back([this, i] { this->run(i); });
        }
    }

    ~PinnedPool() { shutdown(); }

    void submit(int worker, std::function<void()> task) {
        {
            std::lock_guard<std::mutex> lock(mutexes_[worker]);
            queues_[worker].push(std::move(task));
        }
        cvs_[worker].notify_one();
    }

    void shutdown() {
        if (stopping_.exchange(true)) return;
        for (size_t i = 0; i < cvs_.size(); ++i) cvs_[i].notify_all();
        for (auto& t : threads_) {
            if (t.joinable()) t.join();
        }
    }

private:
    void run(int index) {
        for (;;) {
            std::function<void()> task;
            {
                std::unique_lock<std::mutex> lock(mutexes_[index]);
                cvs_[index].wait(lock, [&] { return stopping_ || !queues_[index].empty(); });
                if (queues_[index].empty()) {
                    if (stopping_) return;
                    continue;
                }
                task = std::move(queues_[index].front());
                queues_[index].pop();
            }
            task();
        }
    }

    std::vector<std::queue<std::function<void()>>> queues_;
    std::vector<std::mutex> mutexes_;
    std::vector<std::condition_variable> cvs_;
    std::vector<std::thread> threads_;
    std::atomic<bool> stopping_{false};
};

// Run a task on a chosen worker and wait for it. Keeps the demos readable
// without hiding which thread did the work.
static std::string run_on(PinnedPool& pool, int worker, std::function<std::string()> task) {
    std::mutex m;
    std::condition_variable cv;
    bool done = false;
    std::string result;
    pool.submit(worker, [&] {
        std::string r = task();
        {
            std::lock_guard<std::mutex> lock(m);
            result = std::move(r);
            done = true;
        }
        cv.notify_one();
    });
    std::unique_lock<std::mutex> lock(m);
    cv.wait(lock, [&] { return done; });
    return result;
}

// ---------------------------------------------------------------------------
// The two options.
// ---------------------------------------------------------------------------

// Option A: an explicit context, passed as a parameter. The only place the
// compiler can help you.
struct RequestContext {
    Span span;
};

static std::string trace_id_of(const RequestContext* ctx) {
    return ctx ? ctx->span.trace_id : "none";
}

// Option B: a thread_local. Never cleared, because clearing it is the line
// everybody forgets -- that omission IS the bug being demonstrated.
static thread_local std::optional<Span> tls_current;

static std::string tls_trace_id() {
    return tls_current ? tls_current->trace_id : "none";
}

// ---------------------------------------------------------------------------
// Structured logging.
// ---------------------------------------------------------------------------

struct LogRecord {
    std::string msg;
    std::string trace_id;
};

static std::mutex log_mutex;
static std::vector<LogRecord> logs;

static void log_info(const std::string& msg, const std::string& trace_id) {
    std::lock_guard<std::mutex> lock(log_mutex);
    logs.push_back({msg, trace_id == "none" ? "" : trace_id});
}

static const char* report(const std::string& boundary, const std::string& caller,
                          const std::string& naive, const std::string& propagated,
                          const std::string& note) {
    const char* verdict = (naive == caller) ? "preserved" : "lost";
    std::printf("boundary          %s\n", boundary.c_str());
    std::printf("caller trace_id   %s\n", caller.c_str());
    std::printf("callee trace_id   %-32s naive\n", naive.c_str());
    std::printf("callee trace_id   %-32s propagated\n", propagated.c_str());
    if (note.empty()) {
        std::printf("verdict           %s\n\n", verdict);
    } else {
        std::printf("verdict           %s   (%s)\n\n", verdict, note.c_str());
    }
    return verdict;
}

// ---------------------------------------------------------------------------
// RUN 1: the explicit context. Naive = the handler calls the worker without
// passing ctx (it compiles, because the parameter is a pointer with a default).
// ---------------------------------------------------------------------------

static std::string price_lookup(const RequestContext* ctx, const char* label) {
    log_info(std::string("pricing call (") + label + ")", trace_id_of(ctx));
    return trace_id_of(ctx);
}

static const char* run_explicit_context(PinnedPool& pool) {
    Span span = make_span("GET /orders");
    RequestContext ctx{span};

    // Naive: the lambda captures nothing but the label. There is no context on
    // that thread and no context in the call, so there is no context.
    std::string naive = run_on(pool, 0, [] { return price_lookup(nullptr, "naive"); });

    // Propagated: capture the context by value. A copy per task is the price,
    // and for a struct this size it is not a price.
    RequestContext copy = ctx;
    std::string propagated =
        run_on(pool, 0, [copy]() mutable { return price_lookup(&copy, "ctx passed"); });

    return report("thread pool, explicit context", span.trace_id, naive, propagated,
                  "the compiler cannot warn you: nullptr is a valid argument");
}

// ---------------------------------------------------------------------------
// RUN 2: the thread_local, on a reused thread. The interesting one.
//
// Three requests, two worker threads, dispatched deterministically:
//
//   request A -> worker 0   handler sets tls_current  (normal path)
//   request B -> worker 1   handler sets tls_current  (normal path)
//   request C -> worker 0   handler does NOT set it   (queue-drain path that
//                           assumes "context is already there")
//
// C reports A's trace ID. Not empty. Not an error. A's.
// ---------------------------------------------------------------------------

struct Observation {
    std::string request;
    int worker;
    std::string observed;
    std::string truth;
    bool set_tls;
};

static const char* run_thread_local(PinnedPool& pool) {
    std::vector<Observation> observed;

    auto serve = [&](const char* name, int worker, bool set_tls) {
        Span span = make_span(name);
        std::string truth = span.trace_id;
        std::string seen = run_on(pool, worker, [span, set_tls]() -> std::string {
            if (set_tls) {
                tls_current = span;  // and nobody ever clears it
            }
            log_info(std::string("handling ") + span.name, tls_trace_id());
            return tls_trace_id();
        });
        observed.push_back({name, worker, seen, truth, set_tls});
    };

    serve("req-A", 0, true);
    serve("req-B", 1, true);
    serve("req-C", 0, false);  // lands on A's thread, sets nothing

    const Observation& a = observed[0];
    const Observation& c = observed[2];

    std::printf("boundary          thread_local on a reused pool thread\n");
    std::printf("caller trace_id   %s   (req-C's real trace)\n", c.truth.c_str());
    std::printf("callee trace_id   %-32s naive\n", c.observed.c_str());
    std::printf("callee trace_id   %-32s propagated (explicit ctx, run 1)\n",
                c.truth.c_str());
    const char* verdict = "preserved";
    if (c.observed == c.truth) {
        verdict = "preserved";
    } else if (c.observed == "none") {
        verdict = "lost";
    } else {
        verdict = "WRONG (inherited from previous request)";
    }
    std::printf("verdict           %s\n\n", verdict);

    std::printf("  request  worker  set tls?  observed trace_id                 real trace_id\n");
    for (const auto& o : observed) {
        std::printf("  %-8s %-7d %-9s %-33s %s%s\n", o.request.c_str(), o.worker,
                    o.set_tls ? "yes" : "no", o.observed.c_str(), o.truth.c_str(),
                    (o.observed == o.truth) ? "" : "   <-- MISMATCH");
    }
    std::printf("\n");
    if (c.observed == a.truth) {
        std::printf("  req-C reported req-A's trace ID. Both requests were real, both\n");
        std::printf("  traces exist, and every span of req-C is now filed under req-A.\n");
        std::printf("  A truncated trace makes you hunt for the break. This one does not\n");
        std::printf("  look broken at all -- which is why it is the worse failure.\n\n");
    }
    return verdict;
}

// ---------------------------------------------------------------------------
// RUN 3: a queue -- the message body is the only thing that crosses.
// ---------------------------------------------------------------------------

static const char* run_queue() {
    Span span = make_span("POST /orders");

    struct Message {
        std::string id;
        std::string traceparent;
    };

    auto consume = [](const Message& m) -> std::string {
        // A separate process. Starts from nothing.
        std::optional<Span> restored;
        if (!m.traceparent.empty()) restored = span_from_traceparent(m.traceparent, "job");
        std::string id = restored ? restored->trace_id : "none";
        log_info("processing job " + m.id, id);
        return id;
    };

    std::string naive = consume({"naive", ""});
    std::string propagated = consume({"propagated", traceparent_of(span)});

    return report("Postgres-backed queue", span.trace_id, naive, propagated,
                  "the transport carries no headers; put traceparent in the body");
}

// ---------------------------------------------------------------------------
// RUN 4: the outbound HTTP call -- the easy half, made concrete.
// ---------------------------------------------------------------------------

static const char* run_http() {
    Span span = make_span("GET /orders");
    std::string header = traceparent_of(span);
    auto downstream = span_from_traceparent(header, "GET /price");
    std::printf("boundary          HTTP request to pricing\n");
    std::printf("caller trace_id   %s\n", span.trace_id.c_str());
    std::printf("traceparent sent  %s\n", header.c_str());
    std::printf("callee trace_id   %-32s parsed from the header\n",
                downstream ? downstream->trace_id.c_str() : "PARSE FAILED");
    std::printf("verdict           preserved   (this is what being a W3C standard buys)\n\n");
    return "preserved";
}

int main() {
    std::printf("Layer 6 Topic 3 - losing trace context in C++ (no ambient context at all)\n");
    std::printf("hardware_concurrency=%u   pool: 2 workers, caller-pinned\n",
                std::thread::hardware_concurrency());
    for (int i = 0; i < 72; ++i) std::printf("=");
    std::printf("\n\n");

    PinnedPool pool(2);

    struct Row {
        const char* name;
        const char* verdict;
        const char* who;
    };
    std::vector<Row> rows;
    rows.push_back({"explicit ctx parameter", run_explicit_context(pool), "YOU carry it - by parameter"});
    rows.push_back({"thread_local + reuse", run_thread_local(pool), "YOU carry it - and it lies"});
    rows.push_back({"Postgres queue", run_queue(), "YOU carry it - in the message body"});
    rows.push_back({"http traceparent", run_http(), "the wire format carries it"});

    pool.shutdown();

    std::printf("--- Summary: C++ has no runtime to blame ---\n");
    for (const auto& r : rows) {
        std::printf("  %-26s %-40s %s\n", r.name, r.verdict, r.who);
    }
    std::printf("\n");
    std::printf("  The choice is between a parameter you might forget to pass (loud:\n");
    std::printf("  it is visible at the call site) and a thread_local you might forget\n");
    std::printf("  to set (silent: it reports the last request that used the thread).\n");
    std::printf("  Every other language in this topic makes the same choice; C++ is\n");
    std::printf("  just the one that does not hide either option behind a keyword.\n\n");

    std::printf("--- The one-query test, on the log lines this run emitted ---\n");
    size_t with_id = 0;
    for (const auto& r : logs) {
        if (!r.trace_id.empty()) ++with_id;
    }
    std::printf("  log lines emitted            %zu\n", logs.size());
    std::printf("  lines carrying a trace_id    %zu\n", with_id);
    std::printf("  lines carrying nothing       %zu   <- unqueryable by request\n",
                logs.size() - with_id);
    for (const auto& r : logs) {
        std::printf("    %-28s trace_id=%s\n", r.msg.c_str(),
                    r.trace_id.empty() ? "(empty)" : r.trace_id.c_str());
    }
    std::printf("\n");
    std::printf("  Count the lines that carry an id but the WRONG id: the log pipeline\n");
    std::printf("  cannot tell you which those are, and neither can the trace store.\n");
    return 0;
}

File: 06-observability/03-correlation-ids/golang/lose_the_context.go
// Layer 6 Topic 3 - Losing trace context at a Go concurrency boundary.
//
// What this demonstrates
// ----------------------
// Go refuses to guess. There is no ambient context: `context.Context` is a
// parameter, so nothing is ever lost silently. A function that calls
// `context.Background()` instead of accepting the caller's ctx starts a brand
// new trace -- and that is a bug you can see in a diff, not a mystery at 3am.
//
// So the interesting Go output is not "did it break" (it breaks in exactly the
// place you can point at) but WHERE the break is visible. Each block below
// prints the caller's trace ID, the callee's under the naive call, and the
// callee's when ctx is threaded through. The last column of the summary is the
// one that matters: for every boundary, Go tells you at the call site.
//
// The goroutine block is the one people expect to fail and it does not: a
// goroutine closes over ctx like any other value, so `go work(ctx)` is
// correct by construction. What Go cannot help with is the same thing nothing
// can help with -- a queue, where the context has to be serialised into the
// message body as a W3C `traceparent`.
//
// No OpenTelemetry SDK is imported (none is installed here). The span, the
// context key and the traceparent codec are about 40 lines of standard
// library, which is the point: this is a language property, not an SDK one.
//
// What to look for in the output
// ------------------------------
// Four blocks in the shared shape:
//
//	caller trace_id   <id>
//	callee trace_id   <id or "none">   naive
//	callee trace_id   <id>             propagated
//	verdict           lost | preserved
//
// Then the summary's "fails loudly?" column, which is the whole argument for
// Go's ergonomic trade: every loss here is a visible `context.Background()` or
// `context.TODO()` at a call site, and `go vet`'s lostcancel plus a linter
// like contextcheck will find them mechanically.
package main

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"runtime"
	"strconv"
	"strings"
	"sync"
)

// ---------------------------------------------------------------------------
// A minimal span, a context key, and the W3C traceparent codec.
// ---------------------------------------------------------------------------

type Span struct {
	Name    string
	TraceID string
	SpanID  string
	Sampled bool
}

type ctxKey struct{}

func newSpan(name string) *Span {
	traceID := make([]byte, 16)
	spanID := make([]byte, 8)
	_, _ = rand.Read(traceID)
	_, _ = rand.Read(spanID)
	return &Span{
		Name:    name,
		TraceID: hex.EncodeToString(traceID),
		SpanID:  hex.EncodeToString(spanID),
		Sampled: true,
	}
}

func (s *Span) Traceparent() string {
	flags := "00"
	if s.Sampled {
		flags = "01"
	}
	return fmt.Sprintf("00-%s-%s-%s", s.TraceID, s.SpanID, flags)
}

func spanFromTraceparent(header, name string) (*Span, error) {
	parts := strings.Split(header, "-")
	if len(parts) != 4 || parts[0] != "00" || len(parts[1]) != 32 || len(parts[2]) != 16 {
		return nil, fmt.Errorf("malformed traceparent: %q", header)
	}
	flags, err := strconv.ParseInt(parts[3], 16, 32)
	if err != nil {
		return nil, err
	}
	span := newSpan(name)
	span.TraceID = parts[1]
	span.Sampled = flags&1 == 1
	return span, nil
}

func withSpan(ctx context.Context, s *Span) context.Context {
	return context.WithValue(ctx, ctxKey{}, s)
}

// traceIDOf is the only reader in the file. It returns "none" for a context
// that never carried a span -- which is what context.Background() always is.
func traceIDOf(ctx context.Context) string {
	if s, ok := ctx.Value(ctxKey{}).(*Span); ok {
		return s.TraceID
	}
	return "none"
}

// ---------------------------------------------------------------------------
// Structured logging: the log call takes ctx, because in Go everything takes
// ctx. slog's own context-aware form (slog.InfoContext) exists for exactly
// this reason.
// ---------------------------------------------------------------------------

type logRecord struct {
	msg     string
	traceID string
}

var logs []logRecord

func logInfo(ctx context.Context, msg string) {
	id := traceIDOf(ctx)
	if id == "none" {
		id = ""
	}
	logs = append(logs, logRecord{msg: msg, traceID: id})
}

func report(boundary, caller, naive, propagated, note string) string {
	verdict := "lost"
	if naive == caller {
		verdict = "preserved"
	}
	fmt.Printf("boundary          %s\n", boundary)
	fmt.Printf("caller trace_id   %s\n", caller)
	fmt.Printf("callee trace_id   %-32s naive\n", naive)
	fmt.Printf("callee trace_id   %-32s propagated\n", propagated)
	if note != "" {
		fmt.Printf("verdict           %s   (%s)\n\n", verdict, note)
	} else {
		fmt.Printf("verdict           %s\n\n", verdict)
	}
	return verdict
}

// ---------------------------------------------------------------------------
// Boundary 1: a goroutine. The control, and it passes -- a goroutine closes
// over ctx like any other value.
// ---------------------------------------------------------------------------

func boundaryGoroutine() string {
	span := newSpan("GET /orders")
	ctx := withSpan(context.Background(), span)

	var wg sync.WaitGroup
	var observed string
	wg.Add(1)
	go func(ctx context.Context) {
		defer wg.Done()
		observed = traceIDOf(ctx)
	}(ctx)
	wg.Wait()

	return report("go func(ctx)", span.TraceID, observed, observed,
		"ctx is a value; a goroutine capturing it is correct by construction")
}

// ---------------------------------------------------------------------------
// Boundary 2: the actual Go bug. A helper that manufactures its own context
// because it was easier than changing the signature.
// ---------------------------------------------------------------------------

// fetchPricingNaive is the bug, and it is one identifier long. In a review this
// reads as "it needs a context, here is a context".
func fetchPricingNaive() string {
	ctx := context.Background() // <-- the whole defect
	logInfo(ctx, "GET pricing (naive)")
	return traceIDOf(ctx)
}

func fetchPricing(ctx context.Context) string {
	logInfo(ctx, "GET pricing (ctx threaded)")
	return traceIDOf(ctx)
}

func boundaryFreshContext() string {
	span := newSpan("GET /orders")
	ctx := withSpan(context.Background(), span)

	naive := fetchPricingNaive()
	propagated := fetchPricing(ctx)

	return report("helper calls context.Background()", span.TraceID, naive, propagated,
		"the fix is a parameter; the bug is greppable")
}

// ---------------------------------------------------------------------------
// Boundary 3: a worker pool fed by a channel. Whether context survives depends
// entirely on whether the channel's element type has room for it.
// ---------------------------------------------------------------------------

type jobNoCtx struct {
	id string
}

type jobWithCtx struct {
	id  string
	ctx context.Context
}

func boundaryChannel() string {
	span := newSpan("POST /orders")
	ctx := withSpan(context.Background(), span)

	// Naive: the job struct has no room for a context, so there is nowhere for
	// it to go. The worker calls context.Background() because it must.
	naiveCh := make(chan jobNoCtx, 1)
	naiveOut := make(chan string, 1)
	go func() {
		job := <-naiveCh
		_ = job
		naiveOut <- traceIDOf(context.Background())
	}()
	naiveCh <- jobNoCtx{id: "naive"}
	naive := <-naiveOut

	// Propagated: the job carries the caller's ctx. Go vet will not complain
	// about a context in a struct, but the community convention against it is
	// exactly why in-process queues lose context so often.
	goodCh := make(chan jobWithCtx, 1)
	goodOut := make(chan string, 1)
	go func() {
		job := <-goodCh
		goodOut <- traceIDOf(job.ctx)
	}()
	goodCh <- jobWithCtx{id: "propagated", ctx: ctx}
	propagated := <-goodOut

	return report("chan of jobs -> worker pool", span.TraceID, naive, propagated,
		"in-process: put ctx in the job; cross-process: serialise traceparent")
}

// ---------------------------------------------------------------------------
// Boundary 4: a Postgres-backed queue -- a different process reads a row, so
// only the message body crosses.
// ---------------------------------------------------------------------------

func boundaryQueue() string {
	span := newSpan("POST /orders")
	ctx := withSpan(context.Background(), span)

	type message struct {
		id          string
		traceparent string
	}

	consume := func(m message) string {
		// This runs in `worker`, a separate process. It starts from nothing.
		wctx := context.Background()
		if m.traceparent != "" {
			if s, err := spanFromTraceparent(m.traceparent, "job"); err == nil {
				wctx = withSpan(wctx, s)
			}
		}
		logInfo(wctx, "processing job "+m.id)
		return traceIDOf(wctx)
	}

	naive := consume(message{id: "naive"})
	propagated := consume(message{id: "propagated", traceparent: span.Traceparent()})
	_ = ctx

	return report("Postgres-backed queue (cross-process)", span.TraceID, naive, propagated,
		"the transport carries no headers; put traceparent in the body")
}

// ---------------------------------------------------------------------------
// Boundary 5: the outbound HTTP call -- the easy half, made concrete.
// ---------------------------------------------------------------------------

func boundaryHTTP() string {
	span := newSpan("GET /orders")
	header := span.Traceparent()
	downstream, err := spanFromTraceparent(header, "GET /price")
	if err != nil {
		panic(err)
	}
	fmt.Printf("boundary          HTTP request to pricing\n")
	fmt.Printf("caller trace_id   %s\n", span.TraceID)
	fmt.Printf("traceparent sent  %s\n", header)
	fmt.Printf("callee trace_id   %-32s parsed from the header\n", downstream.TraceID)
	fmt.Printf("verdict           preserved   (this is what being a W3C standard buys)\n\n")
	return "preserved"
}

func main() {
	fmt.Println("Layer 6 Topic 3 - losing trace context in Go (explicit context.Context)")
	fmt.Printf("go %s   %s/%s\n", runtime.Version(), runtime.GOOS, runtime.GOARCH)
	fmt.Println(strings.Repeat("=", 72))
	fmt.Println()

	type row struct {
		name, verdict, loud string
	}
	rows := []row{
		{"go func(ctx)", boundaryGoroutine(), "n/a - does not lose it"},
		{"context.Background() helper", boundaryFreshContext(), "yes - visible at the call site"},
		{"chan job -> worker", boundaryChannel(), "yes - the struct has no ctx field"},
		{"Postgres queue", boundaryQueue(), "no - a missing body field looks like nothing"},
		{"http traceparent", boundaryHTTP(), "n/a - does not lose it"},
	}

	fmt.Println("--- Summary: Go's trade, stated as a table ---")
	fmt.Printf("  %-28s %-10s %s\n", "boundary", "verdict", "fails loudly?")
	for _, r := range rows {
		fmt.Printf("  %-28s %-10s %s\n", r.name, r.verdict, r.loud)
	}
	fmt.Println()
	fmt.Println("  Go loses context in exactly the places you can point at, and the")
	fmt.Println("  one place it cannot help with is the one place no runtime can: a")
	fmt.Println("  queue, where the only thing that crosses is the message body.")
	fmt.Println()

	fmt.Println("--- The one-query test, on the log lines this run emitted ---")
	withID := 0
	for _, r := range logs {
		if r.traceID != "" {
			withID++
		}
	}
	fmt.Printf("  log lines emitted            %d\n", len(logs))
	fmt.Printf("  lines carrying a trace_id    %d\n", withID)
	fmt.Printf("  lines carrying nothing       %d   <- unqueryable by request\n", len(logs)-withID)
	for _, r := range logs {
		id := r.traceID
		if id == "" {
			id = "(empty)"
		}
		fmt.Printf("    %-28s trace_id=%s\n", r.msg, id)
	}
	fmt.Println()
	fmt.Println("  Every one of those calls took a ctx. The empty lines are the ones")
	fmt.Println("  where the ctx handed in was manufactured rather than passed down.")
}
